File: euler-201-subsets-with-a-unique-sum.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func main() {
	args := os.Args[1:] // Get the arguments
	nums := make([]int, len(args))
	for i, arg := range args { // Creates a slice with all arguments
		num, _ := strconv.Atoi(arg)
		nums[i] = num
	}
	subset_size := 3
	combinations := getcombinations(nums, len(nums), subset_size)
	fmt.Println(get_unique_subset_sum(combinations)) // Print sum of unique subset sums
}

// Uses the generate_subsets function to store all subset combinations
func getcombinations(nums []int, length int, subset_size int) [][]int {
	curr := make([]int, subset_size)
	combinations := [][]int{}
	generate_subsets(&combinations, nums, curr, 0, length-1, 0, subset_size) // An address to combinations allows generate_subsets to make changes to combinations
	return combinations
}

//Algorithm to generate all subsets of a given size from a set of numbers
//Made with the help of https://www.geeksforgeeks.org/print-all-possible-combinations-of-r-elements-in-a-given-array-of-size-n/
func generate_subsets(combinations *[][]int, nums []int, curr []int, start int, end int, index int, subset_size int) {

	if index == subset_size { //Checks if curr is filled
		temp := make([]int, subset_size)
		copy(temp, curr)
		*combinations = append(*combinations, temp) // I have no idea why I need to make a temp variable for this but it only works like this
		return
	}

	// Loops from the numbers i to end and ensures that the number of remaining numbers in the set is greater than or equal to the subset size
	for i := start; i <= end && end - i + 1 >= subset_size - index; i++ {
		curr[index] = nums[i]
		generate_subsets(combinations, nums, curr, i + 1, end, index + 1, subset_size); //Recursive call of generate_subsets
	}
}

// Finds the sum of all unique subset sums
func get_unique_subset_sum(subsets [][]int) int {
	sum_counter := make(map[int]int) // Creates a map that holds sums as keys and how many times they occur in their value
	sum := 0
	for _, subset := range subsets { // Iterate through all subsets

		for _, num := range subset { // Iterate through the numbers in the subset
			sum += num // Sum up the subset numbers
		}

		// Check if sum_counter contains sum, ref: https://golang.cafe/blog/how-to-check-if-a-map-contains-a-key-in-go
		_, ok := sum_counter[sum]
		if ok {
			sum_counter[sum] += 1
		} else {
			sum_counter[sum] = 1
		}
		sum = 0
	}
	sum_of_unique_sums := 0
	for key, value := range sum_counter {
		if value == 1 {
			sum_of_unique_sums += key // Adds key to sum of unique sum if the sum occured just once
		}
	}

	return sum_of_unique_sums
}

File: euler_201_subsets_with_a_unique_sum.java
import java.util.*;

public class euler_201_subsets_with_a_unique_sum {
    public static void main(String[] args) {
        List<Integer> nums = new ArrayList<>(); //New integer array

        //Add all arguments to the array, creating our set
        for (String arg : args) {
            nums.add(Integer.parseInt(arg));
        }
        int subsetSize = 3;
        List<List<Integer>> combinations = getCombinations(nums, nums.size(), subsetSize); //Generate combinations
        System.out.println(getUniqueSubsetSum(combinations)); //Print the sum of all unique subset sums
    }

    //Uses generate subsets method to get all combinations
    public static List<List<Integer>> getCombinations(List<Integer> nums, int length, int subsetSize) {
        List<Integer> curr = new ArrayList<>();

        //Fill curr's indexes to avoid indexing errors
        for (int i = 0; i < subsetSize; i++) {
            curr.add(0);
        }
        List<List<Integer>> combinations = new ArrayList<>();
        generateSubsets(combinations, nums, curr, 0, length - 1, 0, subsetSize);
        return combinations;
    }

    //Generate all combinations of a given size of a set of numbers
    public static void generateSubsets(List<List<Integer>> combinations, List<Integer> nums, List<Integer> curr, int start, int end, int index, int subsetSize) {

        //Check if a new combination is ready
        if (index == subsetSize) {
            combinations.add(new ArrayList<>(curr));
            return;
        }
 
        //Loop from start to end of the set of numbers and ensure that there always remains enough numbers in the set to generate a subset
        for (int i = start; i <= end && end - i + 1 >= subsetSize - index; i++) {
            curr.set(index, nums.get(i));
            generateSubsets(combinations, nums, curr, i + 1, end, index + 1, subsetSize); //Recursive call
        }
    }

    //Calculates and returns the sum of all unique subset sums
    public static int getUniqueSubsetSum(List<List<Integer>> subsets) {
        Map<Integer, Integer> sumCounter = new HashMap<>(); //Create a new map of integer keys and values
        int sum = 0;

        //For every subset in subsets
        for (List<Integer> subset : subsets) {

            //For every number in the subset
            for (int num : subset) {
                sum += num;
            }

            //If the sum already exists in the map, increment the value, otherwise add the sum with a value of 1
            if (sumCounter.containsKey(sum)) {
                sumCounter.put(sum, sumCounter.get(sum) + 1);
            } else {
                sumCounter.put(sum, 1);
            }
            sum = 0;
        }
        int sumOfUniqueSums = 0;

        // Loop through hashmap, ref: https://stackoverflow.com/questions/1066589/iterate-through-a-hashmap
        for (Map.Entry<Integer, Integer> mapEntry : sumCounter.entrySet()) {

            //If the counter of the sum is 1 (is it a unique sum?)
            if (mapEntry.getValue() == 1) {
                sumOfUniqueSums += mapEntry.getKey();
            }
        }
        return sumOfUniqueSums;
    }
}